compiler/rustc_hir_typeck/src/closure.rs

```diff
@@ -650,9 +650,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         },
                     )
                 }
-                // For a `gen {}` block created as a `gen fn` body, we need the return type to be
-                // ().
-                Some(hir::CoroutineKind::Gen(hir::CoroutineSource::Fn)) => self.tcx.types.unit,
+                // All `gen {}` and `async gen {}` must return unit.
+                Some(hir::CoroutineKind::Gen(_) | hir::CoroutineKind::AsyncGen(_)) => {
```

Comment:
RFC 3513 [says](https://github.com/rust-lang/rfcs/pull/3513/files#diff-e764793d5d58a879a888ed6970bd66d89f5b79e00c30480bc21038fdde5116ecR110) gen blocks either return `()` or `!`. Do we need to account for diverging return types here or is that just not implemented at this time?

Reply:
There shouldn't be a problem with always making gen blocks return `()`, since that return type is neither accessible, and also `!` will always coerce to `()`. I'll add a test.

Reply:
Done